src/RLInterface.jl

```diff
@@ -29,72 +29,86 @@ export
 
 abstract type AbstractEnvironment end
 
-mutable struct MDPEnvironment{S} <: AbstractEnvironment
-    problem::MDP
+"""
+    obsvector_type(::Union{MDP, POMDP})
+
+Returns the type of the observation vector associated with a specific problem. 
+The `MDPEnvironment` and `POMDPEnvironment` wrappers will convert observations to an object of such type when `reset!` or `step!` is called.
+"""
+obsvector_type(::Union{MDP, POMDP}) = Vector{Float32}
+
+mutable struct MDPEnvironment{OV, M<:MDP, S, R<:AbstractRNG} <: AbstractEnvironment 
+    problem::M
     state::S
-    rng::AbstractRNG
+    rng::R
 end
-function MDPEnvironment(problem::MDP; rng::AbstractRNG=MersenneTwister(0))
-    return MDPEnvironment(problem, initialstate(problem, rng), rng)
+function MDPEnvironment(problem::M,
+                        ov::Type{A} = obsvector_type(problem);
+                        rng::R=MersenneTwister(0)) where {A<:AbstractArray, M<:MDP, R<:AbstractRNG}
+    S = statetype(problem)
+    return MDPEnvironment{ov, M, S, R}(problem, initialstate(problem, rng), rng)
 end
 
-mutable struct POMDPEnvironment{S} <: AbstractEnvironment
-    problem::POMDP
+mutable struct POMDPEnvironment{OV, M<:POMDP, S, R<:AbstractRNG} <: AbstractEnvironment
+    problem::M
     state::S
-    rng::AbstractRNG
+    rng::R
 end
-function POMDPEnvironment(problem::POMDP; rng::AbstractRNG=MersenneTwister(0))
-    return POMDPEnvironment(problem, initialstate(problem, rng), rng)
+function POMDPEnvironment(problem::M,
+                          ov::Type{A} = obsvector_type(problem);
+                          rng::R=MersenneTwister(0)) where {A<:AbstractArray, M<:POMDP, R<:AbstractRNG}
+    S = statetype(problem)
+    return POMDPEnvironment{ov, M, S, R}(problem, initialstate(problem, rng), rng)
 end
 
 """
-    reset!(env::MDPEnvironment)
+    reset!(env::MDPEnvironment{OV}) where OV
 Reset an MDP environment by sampling an initial state returning it.
 """
-function reset!(env::MDPEnvironment)
+function reset!(env::MDPEnvironment{OV}) where OV
     s = initialstate(env.problem, env.rng)
     env.state = s
-    return convert_s(Array{Float64, 1}, s, env.problem)
+    return convert_s(OV, s, env.problem)
 end
 
 """
-    reset!(env::POMDPEnvironment)
+    reset!(env::POMDPEnvironment{OV}) where OV
 Reset an POMDP environment by sampling an initial state,
 generating an observation and returning it.
 """
-function reset!(env::POMDPEnvironment)
+function reset!(env::POMDPEnvironment{OV}) where OV
     s = initialstate(env.problem, env.rng)
     env.state = s
     a = first(actions(env))
     o = generate_o(env.problem, s, a, s, env.rng)
-    return convert_o(Array{Float64, 1}, o, env.problem)
+    return convert_o(OV, o, env.problem)
 end
 
 """
-    step!{A}(env::POMDPEnvironment, a::A)
+    step!{A}(env::MDPEnvironment{OV}, a::A) where {A, OV}
```

Comment:
I'd recommend just getting rid of all parameters in docstrings. Definitely remove {A} before the arguments

Reply:
There are other places where this should be done below too

Reply:
I always add them just to match the signature, is there a specific reason for this convention?

Reply:
Does it hold for the struct docstrings? e.g. : 
`KMarkovEnvironment{OV, M<:POMDP, S, R<:AbstractRNG} <: AbstractEnvironment`

should become
`KMarkovEnvironment`

Reply:
> I always add them just to match the signature, is there a specific reason for this convention?

I just think it's easier to read, but that's personal preference. Up to you whether to keep it. The main thing I wanted to point out is the `{A}` in `step!{A}(env...`.

I think I would keep them for the struct docstrings.

Reply:
ah! got it, should be good now.